akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -671,20 +591,27 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
   protected def mergeAttributes(parent: Attributes, current: Attributes): Attributes =
     parent and current
 
+  private val matValSrc: ju.Map[MaterializedValueNode, List[MaterializedValueSource[Any]]] = new ju.HashMap
+  def registerSrc(ms: MaterializedValueSource[Any]): Unit = {
+    if (MaterializerSession.Debug) println(s"registering source $ms")
+    matValSrc.get(ms.computation) match {
+      case null ⇒ matValSrc.put(ms.computation, ms :: Nil)
+      case xs   ⇒ matValSrc.put(ms.computation, ms :: xs)
+    }
+  }
+
   protected def materializeModule(module: Module, effectiveAttributes: Attributes): Any = {
-    val materializedValues = collection.mutable.HashMap.empty[Module, Any]
-    var materializedValuePublishers: List[MaterializedValuePublisher] = Nil
+    val materializedValues: ju.Map[Module, Any] = new ju.HashMap
 
     for (submodule ← module.subModules) {
       val subEffectiveAttributes = mergeAttributes(effectiveAttributes, submodule.attributes)
       submodule match {
-        case mv: MaterializedValueSource[_] ⇒
-          val pub = new MaterializedValuePublisher
-          materializedValuePublishers ::= pub
-          materializedValues.put(mv, ())
-          assignPort(mv.shape.outlet, pub)
+        case GraphStageModule(shape, attributes, mv: MaterializedValueSource[_]) ⇒
+          val copy = mv.copy.asInstanceOf[MaterializedValueSource[Any]]
```

Comment:
every inclusion of a MaterializedValueSource must have its own Promise—they are not reusable